libs/pilight/protocols/433.92/hideki.c

```diff
@@ -0,0 +1,497 @@
+ * The payload (excluding the header) has an byte parity (XOR) check
+ * The payload (excluding the header) has CRC-8, poly 0x07 init 0x00 check
+ * The payload bytes are reflected (LSB first / LSB last) after the CRC check
+ *
+ *    11111001 0  11110101 0  01110011 1 01111010 1  11001100 0  01000011 1  01000110 1  00111111 0  00001001 0  00010111 0
+ *    SYNC+HEAD P   RC cha P     LEN   P     Nr.? P   .1° 1°  P   10°  BV P   1%  10% P     ?     P     XOR   P     CRC   P
+ *
+ * TS04:
+ *    00000000  11111111  22222222  33333333  44444444  55555555  66666666  77777777  88888888 99999999
+ *    SYNC+HEAD cha   RC     LEN        Nr.?    1° .1°  VB   10°   10%  1%     ?         XOR      CRC
+ *
+ * Wind:
+ *    00000000  11111111  22222222  33333333  44444444  55555555  66666666  77777777  88888888 99999999 AAAAAAAA BBBBBBBB CCCCCCCC DDDDDDDD
+ *    SYNC+HEAD cha   RC     LEN        Nr.?    1° .1°  VB   10°    1° .1°  VB   10°   1W .1W  .1G 10W   10G 1G    w°  AA    XOR      CRC
+ *
+ * Rain:
+ *    00000000  11111111  22222222  33333333  44444444  55555555  66666666  77777777  88888888
+ *    SYNC+HEAD cha   RC   B LEN        Nr.?   RAIN_L    RAIN_H     0x66       XOR       CRC
+ *
+ */
+
+enum sensortypes { HIDEKI_UNKNOWN, HIDEKI_TEMP, HIDEKI_TS04, HIDEKI_WIND, HIDEKI_RAIN };
+
+typedef struct settings_t {
+	/* This driver implements different sensor types, that are dynamically detected in parseCode() function.
+	 * Unfortunately, the detected sensor type sometimes is wrong. Unfortunately, the expected sensor type
+	 * as defined in the user configuration ("protocol" name from config.json) is not accessible in function
+	 * checkValues(). Hence a heuristics tries to get the right sensor type and stores them in this settings
+	 * structure:
+	 * In the first received packets, the system stores the sensor type as soon as the same type was detected
+	 * in (CALIBRATION_COUNT+1) subsequent data packets. From this time on, all decoded packets with
+	 * wrong sensor type are ignored */
+	uint8_t calibration;
+	uint8_t sensortype;
+	uint8_t ignore_temp;
+
+	double rc; // double: this type seems necessary because of the data types expected by JSON
+	double channel;
+	double temp;
+	double humi;
+	double wind_dir;
+	double wind_factor;
+	double rain;
+	struct settings_t *next;
+} settings_t;
+
+static struct settings_t *settings = NULL;
+
+static uint8_t reverse8(uint8_t x) {
+	x = (x & 0xF0) >> 4 | (x & 0x0F) << 4;
+	x = (x & 0xCC) >> 2 | (x & 0x33) << 2;
+	x = (x & 0xAA) >> 1 | (x & 0x55) << 1;
+	return x;
+}
+
+static int byteParity(uint8_t inByte) {
+	inByte ^= inByte >> 4;
+	inByte &= 0xf;
+	return (0x6996 >> inByte) & 1;
+}
+
+static void set_bit(uint8_t *array, int position) {
+	uint8_t mask = 1 << (7 - (position % 8));
+	array[position / 8] |= mask;
+}
+
+static struct settings_t *get_settings(double channel) {
+	struct settings_t *tmp = settings;
+
+	while(tmp) {
+		if(fabs(tmp->channel - channel) < EPSILON)  {
```

Comment:
Isn't it easier to cast channel to an int below so we can just compare the numbers directly without having to rely on the `fabs` and `EPSILON`?

Reply:
I agree with you. As I checked the corresponding code in other pilight protocol implementations I found that this is done in the same way. The I decided to leave it using double.

Reply:
Maybe i was too tired back then. So let's not copy the same wierd implementations over and over.

Reply:
OK, I will think about it.... but not today.